packages/main/src/plugin/kubernetes-context-state.ts

```diff
@@ -0,0 +1,227 @@
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ ***********************************************************************/
+
+import type {
+  Informer,
+  KubernetesObject,
+  ListPromise,
+  ObjectCache,
+  V1Deployment,
+  V1Pod,
+} from '@kubernetes/client-node';
+import { AppsV1Api, CoreV1Api, KubeConfig, makeInformer } from '@kubernetes/client-node';
+import type { KubeContext } from './kubernetes-context.js';
+import type { ApiSenderType } from './api.js';
+
+// ContextInternalState stores informers for a kube context
+interface ContextInternalState {
+  podInformer?: Informer<V1Pod> & ObjectCache<V1Pod>;
+  deploymentInformer?: Informer<V1Deployment> & ObjectCache<V1Deployment>;
+}
+
+// ContextState stores information for the user about a kube context: is the cluster reachable, the number
+// of instances of different resources
+export interface ContextState {
+  reachable: boolean;
+  podsCount: number;
+  deploymentsCount: number;
+}
+
+interface CreateInformerOptions<T> {
+  checkReachable?: boolean;
+  onAdd?: (obj: T) => void;
+  onDelete?: (obj: T) => void;
+  onReachable?: (reachable: boolean) => void;
+}
+
+// the ContextsState singleton (instantiated by the kubernetes-client singleton)
+// manages the state of the different kube contexts
+export class ContextsState {
+  private kubeConfig = new KubeConfig();
+  private contextsState = new Map<string, ContextState>();
+  private contextsInternalState = new Map<string, ContextInternalState>();
+
+  constructor(private readonly apiSender: ApiSenderType) {}
+
+  // update is the reconcile function, it gets as input:
+  // - the user preference indicating if the user wants to get live information about kube contexts
+  // - the last known kube config
+  // and starts/stops informers for different kube contexts, depending on these inputs
+  async update(kubeconfig: KubeConfig) {
+    this.kubeConfig = kubeconfig;
+    // Add informers for new contexts
+    for (const context of this.kubeConfig.contexts) {
+      if (this.contextsInternalState.get(context.name) === undefined) {
+        this.contextsState.set(context.name, {
+          reachable: false,
+          podsCount: 0,
+          deploymentsCount: 0,
+        });
+        const informers = this.createKubeContextInformers(context);
+        if (informers) {
+          this.contextsInternalState.set(context.name, informers);
+        }
+      }
+    }
+    // Delete informers for removed contexts
+    let removed = false;
+    for (const [name, state] of this.contextsInternalState) {
+      if (!this.kubeConfig.contexts.find(c => c.name === name)) {
+        await state.podInformer?.stop();
+        await state.deploymentInformer?.stop();
```

Comment:
should we have only a stop method that does that for us, instead of the need to call sub parts

in addition, It's not also mapping overall arch that is using disposable pattern

so when I register something I have a disposable object that will do stop, delete, etc
cleaning is just calling the dispose() method on the disposable object

so I would have only called dispose() on the state and it would have self stopped/removed the resource

Reply:
it should be fixed with latest commit (refactor: use dispose pattern for state)